Added E2E execution plan tests with coverage

laurennat opened the pull request:
Set up playwright infrastructure for collecting coverage on react webviews and testing react webviews. There's also a few really minor changes to some Execution Plan react components, which just make it easier to playwright test them.

![Screenshot 2025-03-12 124128](https://github.com/user-attachments/assets/397c61ba-e4c3-4bd5-887e-c82c32615611)

caohai commented:
> This is also only for local testing.

Can you elaborate a bit on this?

laurennat commented:
Yes, I'm adding it in later

aasimkhan30 commented on test/e2e/utils/executionPlanHelpers.ts:
```diff
@@ -0,0 +1,169 @@
+}
+
+export async function tabToQueryPlanToolbar(vsCodePage: Page): Promise<void> {
+    // Go to query plan toolbar from main state
```
I don't know if this will be scalable. Should we think about creating mock vscode socket and run these the webview panels in the browser? I think we were doing something for Fabric.

aasimkhan30 commented on test/e2e/executionPlan.spec.ts:
```diff
@@ -0,0 +1,288 @@
+            vsCodePage,
+            "Actual Number of Rows For All Executions",
+        );
+        await checkScreenshot(
```
What does check screenshot functionality do exactly? Since we cannot access contents on the page,  how will we check if our key downs  are having intended effect?

laurennat replied:
The checkScreenshot functionality will compare the current visual state of the extension host (which is where the tests are run) with the expected visual state.

caohai commented on test/e2e/README_webviewTesting.md:
```diff
@@ -0,0 +1,54 @@
+
+For the bulk of the Playwright testing within the react webviews, we test the functionality by comparing screenshots of the expected results to the current test state.
+
+We can't use playwright's builtin locator functions, because VSCode webviews are sandboxed- due to cross-origin restrictions, their HTML content can't be accessed outside the webview context (which unfortunately is where the Playwright test suite is run).
```
Have you looked into https://playwright.dev/docs/api/class-framelocator?
https://github.com/eclipse-theia/theia/discussions/14896#discussioncomment-12179905

laurennat replied:
yes, it's not accessible from the parent context :(

aasimkhan30 commented:
Using the example, I was able to access elements inside the webviews. I opened the connection dialog, found the webview iframe, located the profile name field, and filled it in with a dummy profile name.

```ts
   const connectionDialogFrame = vsCodePage
        .frameLocator(".webview")
        .frameLocator("[title='Connection Dialog (Preview)']");

    if (connectionDialogFrame) {
        const inputField = connectionDialogFrame.getByLabel("Profile Name");
        await expect(inputField).toBeVisible({ timeout: 10000 });
        const isInputFieldVisible = await inputField.isVisible();
        if (isInputFieldVisible) {
            await inputField.fill("Server Name");
        }
    }
```

caohai commented:
> Using the example, I was able to access elements inside the webviews. I opened the connection dialog, found the webview iframe, located the profile name field, and filled it in with a dummy profile name.
> 
> ```ts
>    const connectionDialogFrame = vsCodePage
>         .frameLocator(".webview")
>         .frameLocator("[title='Connection Dialog (Preview)']");
> 
>     if (connectionDialogFrame) {
>         const inputField = connectionDialogFrame.getByLabel("Profile Name");
>         await expect(inputField).toBeVisible({ timeout: 10000 });
>         const isInputFieldVisible = await inputField.isVisible();
>         if (isInputFieldVisible) {
>             await inputField.fill("Server Name");
>         }
>     }
> ```

This looks great!

laurennat commented:
> > Using the example, I was able to access elements inside the webviews. I opened the connection dialog, found the webview iframe, located the profile name field, and filled it in with a dummy profile name.
> > ```ts
> >    const connectionDialogFrame = vsCodePage
> >         .frameLocator(".webview")
> >         .frameLocator("[title='Connection Dialog (Preview)']");
> > 
> >     if (connectionDialogFrame) {
> >         const inputField = connectionDialogFrame.getByLabel("Profile Name");
> >         await expect(inputField).toBeVisible({ timeout: 10000 });
> >         const isInputFieldVisible = await inputField.isVisible();
> >         if (isInputFieldVisible) {
> >             await inputField.fill("Server Name");
> >         }
> >     }
> > ```
> 
> This looks great!

I'll change the tests to this format! thanks for finding this

aasimkhan30 commented on test/e2e/executionPlan.spec.ts:
```diff
@@ -0,0 +1,429 @@
+        // Press Enter in the VS Code page
+        await vsCodePage.keyboard.press("Enter");
+
+        iframe = vsCodePage
```
Wrap this in a helper function.

aasimkhan30 commented on .github/workflows/daily-build-and-test.yml:
The status of this run is used in Readme. Please fix that too.

aasimkhan30 commented on .github/workflows/pr-checks.yml:
```diff
@@ -6,9 +6,15 @@ name: PR Checks
+  push:
+    branches:
+      - main
   pull_request:
```
Lets add it to pull requests in release branch as well.

aasimkhan30 commented on .gitignore:
```diff
@@ -15,6 +15,7 @@ tools
 examples
 sqltoolsservice
 coverage
+coverage-e2e
```
Maybe just put the coverage inside the coverage folder to keep the root dirs cleaner.

aasimkhan30 commented on .gitignore:
```diff
@@ -29,4 +30,7 @@ package.nls.*.json
\ No newline at end of file
+.eslintcache
+.nyc_output
+/test/resources/screenshots
```
maybe it should be test/artifacts. Since resource imply external files used as inputs in tests.

aasimkhan30 commented on .vscodeignore:
```diff
@@ -26,4 +27,7 @@ tslint*.json
\ No newline at end of file
+FEATURES.md
+.nyc_output
+/test/resources/screenshots
```
test folder is ignored so this isn't need.

aasimkhan30 commented on .gitignore:
```diff
@@ -29,4 +30,7 @@ package.nls.*.json
+.eslintcache
+.nyc_output
+/test/resources/screenshots
+reactviews-coverage
```
let's try to put everything inside one coverage dir at root level.

aasimkhan30 commented on gulpfile.js:
```diff
@@ -10,7 +10,7 @@ const minifier = require('gulp-uglify/minifier');
 const argv = require('yargs').argv;
 const min = (argv.min === undefined) ? false : true;
-const prod = (argv.prod === undefined) ? false : true;
+const prod = (argv.prod === "false" || argv.prod === undefined) ? false : true;
```
```
const prod = argv.prod === "true"; 
```

aasimkhan30 commented on package.json:
```diff
@@ -55,14 +55,17 @@
     "clean-package": "git clean -xfd && yarn install && yarn build && yarn gulp package:online",
-    "testWithCoverage": "yarn test && yarn gulp cover"
+    "testWithCoverage": "yarn test && yarn gulp cover",
+    "instrument": "nyc instrument ./out/src ./out/src --in-place --exclude=**/views/htmlcontent/**",
```
Maybe combine these commands into

aasimkhan30 commented on test/resources/mergeReports.js:
```diff
@@ -0,0 +1,93 @@
+const fs = require("fs");
```
Can you use something like this?

https://www.npmjs.com/package/@jtmdias/merge-coverage
https://www.npmjs.com/package/cobertura-merge

laurennat replied:
The generated report for the React views is placed under the package name main instead of src.reactviews.pages, so we’ll need to update that before merging the reports. Also, I tried the other two packages you recommended, but the default formatting for the merged reports is a bit off. we’d probably need to adjust the XML files manually to get the desired structure anyways

Benjin replied:
Just to know, how "off" are the two formats?  A custom XML merger isn't something I expected to find in this PR.

aasimkhan30 commented on test/e2e/utils/testHelpers.ts:
```diff
@@ -119,3 +116,5 @@ export async function waitForCommandPaletteToBeVisible(
     await expect(commandPaletteInput).toBeVisible();
 }
+
+export let numScreenshotsForTest = new Map<string, number>();
```
is this still used?

aasimkhan30 commented on test/e2e/utils/launchVscodeWithMsSqlExt.ts:
```diff
@@ -37,5 +37,8 @@ export async function launchVsCodeWithMssqlExtension(): Promise<{
     });
 
+    // wait for 5 seconds for the extension to load
+    await new Promise((resolve) => setTimeout(resolve, 5 * 1000));
```
Can we use playwright to determine if the extensions has loaded? Something like checking OE contents or something? Timeouts can be very flaky

aasimkhan30 replied:
If it is not possible. It might be helpful to add a prompt that appears when the service is loaded during test runs. This would give us a deterministic way to confirm that the extension has fully initialized before we start executing tests.

@Benjin and @caohai any other ideas?

laurennat commented on test/e2e/README_webviewTesting.md:
````diff
@@ -0,0 +1,41 @@
+
+After the code is instrumented run the tests using
+```shell
+npx nyc --reporter=html --reporter=text-summary --include="src/reactviews/pages/**/*.tsx" npx playwright test
````
instrumenting does take a while

Benjin commented on test/e2e/executionPlan.spec.ts:
```diff
@@ -0,0 +1,430 @@
+/*---------------------------------------------------------------------------------------------
```
Thanks for adding these! 🙌

Benjin commented on test/resources/plan.sqlplan:
```diff
@@ -0,0 +1 @@
+<ShowPlanXML xmlns="http://schemas.microsoft.com/sqlserver/2004/07/showplan" Version="1.564" Build="16.0.4135.4"><BatchSequence><Batch><Statements><StmtSimple StatementText="select * from sys.all_views" StatementId="1" StatementCompId="1" StatementType="SELECT" RetrievedFromCache="false" StatementSubTreeCost="0.219688" StatementEstRows="763.17" SecurityPolicyApplied="false" StatementOptmLevel="FU ...
```
Can we format this XML into something human-readable before checking it in?  It'd make future diffs easier.

Benjin commented:
![image](https://github.com/user-attachments/assets/1fe35393-4702-43b5-8e12-641a5ba02f34)

Do you know what's going on here?

laurennat commented:
> ![image](https://private-user-images.githubusercontent.com/1609827/427145512-1fe35393-4702-43b5-8e12-641a5ba02f34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMjA1MDYsIm5iZiI6MTc0MzAyMDIwNiwicGF0aCI6Ii8xNjA5ODI3LzQyNzE0NTUxMi0xZmUzNTM5My00NzAyLTQzYjUtOGUxMi02NDFhNWJhMDJmMzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjZUMjAxNjQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWY4Nzg2NzE0NzdiNjMzZDhjYmVjNjRmZTQxMjZmZTU2M2M5YjQ0MjZjOTc4MTM4MmI3YTg2NWZlOTllZDRlMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.j_6vr50Wvnrga36dViG4xCXXSSaYQDl74KDmyg29bkw)
> 
> Do you know what's going on here?

when I changed the prod flag to explicitly have to be set to true, I forgot to set it in the build process in the pipeline. I've fixed that now, but the changes have not been overwritten because the smoke tests are broken.

laurennat commented on .github/workflows/pr-checks.yml:
```diff
@@ -122,6 +125,120 @@ jobs:
+          echo "line_rate_pr=$line_rate_pr" >> $GITHUB_ENV
+          echo "line_rate_diff=$(echo "$line_rate_pr - $line_rate_main" | bc)" >> $GITHUB_ENV
+          coverage_text_color=$(if [ $line_rate_diff -gt 0 ]; then echo "red"; else echo "lightgreen"; fi)
+          echo "coverage_text_color=$coverage_text_color" >> $GITHUB_ENV
```
i did not write this script but sure

laurennat commented on .github/workflows/pr-checks.yml:
```diff
@@ -6,9 +6,16 @@ name: PR Checks
 
 on:
+  schedule:
+    - cron: '0 0 * * *'
```
I merged the CI and PR runs into one, otherwise we'd be doing the same thing in both- the only task unique to the CI run before was the smoke tests.

laurennat replied:
maybe a rename is in store...

laurennat commented:
> ![image](https://private-user-images.githubusercontent.com/1609827/427145512-1fe35393-4702-43b5-8e12-641a5ba02f34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMwMzA4MDYsIm5iZiI6MTc0MzAzMDUwNiwicGF0aCI6Ii8xNjA5ODI3LzQyNzE0NTUxMi0xZmUzNTM5My00NzAyLTQzYjUtOGUxMi02NDFhNWJhMDJmMzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMjZUMjMwODI2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmIzOTY5ZDI2NjIwZjNmZmQyYjFmNGUwNjgwNDdlNGI3YzIwMmZjZGNmZGY2ZjhhOGQ1NjQ1NDk1ZThiZDBkMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.wxqZKLhXoLXZvuY582Sj1ySutFeVXARUr8qo-LlX6qs)
> 
> Do you know what's going on here?

This was fixed, but it couldn't overwrite the results because the pipeline was broken :( Now it's good though!

laurennat commented:
@lewis-sanchez is merging for me because I still have the merge permissions issue
